### ISSUE TITLE
whatshap for polyploid without evenly balanced alleles

### ISSUE THREAD

**Issue by michaeljmetzger** [whatshap/whatshap#409]
We are trying to use WhatsHap to phase data on mixed samples with more than 2 alleles, and they are not all in an equal ratio. For a test sample, we have 3 alleles that are in a roughly 49:49:2 ratio. We can get polyphase to run and give us the expected output (0|0|1) if we specify the known genotypes in advance, but if we use the distrust genotypes option on our test sample, it removed the third allele that has lower support (the SNP is present in 13 reads out of 414 at the SNP in question). We get 0|0|0 for that SNP instead of 0|0|1.  

I am assuming there are parameters that allow the multiple alleles to deviate from a 1:1:1 ratio or maybe a threshold of a certain % coverage needed to call a genotype as real. Is there any way to modify these parameters to allow for the detection of low-frequency alleles in a mix?

Thanks,
Michael

**Comment by schrins** [comment]
As of now, there is no way to promote unevenly balanced phases. The assumption of the phaser is, that the coverage per position is on average evenly distributed among the haplotypes. I also do not see an easy way to use a distribution in the algorithm, because there is no order of the haplotypes at any time.

**Comment by michaeljmetzger** [comment]
I don't think we need to add in the use of any particular distribution or different algorithm. It might be doable just changing existing parameters. I wasn't sure whether whatshap assumed even distributions or had a threshold.

If the phaser assumes an even distribution, I am guessing it doesn't require it to be exact (ie. for a diploid SNP it would expect 50 out of 100 reads to be the heterozygous SNP, but it would still call it heterozygous if it was 49 or 51 reads out of 100). There is probably a parameter which determines how far from 50:50 is allowable or some weight to this measure. If so, is there a way to change either?

If instead of distance from 50%, the phaser uses a threshold (ie. >10% of reads must be the alternate SNP or it is removed), is there a way to alter that parameter?

I know I am trying to use this for something that it was not quite designed for, and thanks for your help and time on this.

**Comment by schrins** [comment]
Sadly, there is no parameter or threshold for uneven haplotype coverage. The way the phaser works is, that it first creates clusters of reads based on similarity and then lays k haplotype threads (k = ploidy) over these clusters, such that for any position p the number of threads over a cluster somewhat reflects its read coverage for position p. The more the coverage distribution induced by the threads deviates from the even distribution, the more the thread configuration is penalized for this position. Thus, short coverage fluctuations do not matter much. With your data, the phaser will probably create read clusters for the lowly coveraged haplotype, but it will not lay any threads over it, because "wasting" a full haplotype thread on such a low coverage cluster will constantly create high penalties and the phaser will rather squeeze the third thread into the clusters of the other two haplotypes. The reason why the you sometimes still got the right genotypes with default parameters is, that after determining the haplotype threads (and their consensus for each position) there is cleaning step, where the all positions are forced to conform to the input genotypes. If the phaser predicted 0/0/0, but the input says 0/0/1, then the phase covering the highest fraction of 1-alleles with its reads, will turn into a 1. In your case, this is not going to be meaningful, the three predicted haplotypes are based on actually two real haplotypes and it will be more or less random, which of the phases will switch to the "missing" allele on every position.

**Comment by michaeljmetzger** [comment]
Thanks so much for the explanation. Then the only parameter that might help would be a parameter involved in weighting the penalties. But I am guessing that lowering the penalties for this process could break the whole phasing step. That is good to know that the original genotypes are just forced onto the data in this way and it wouldn't help up to use an external method to call the genotypes more accurately first.

**Comment by schrins** [comment]
Besides the coverage penalty, there is a second penalty for the threads, namely switches. Every time a thread switches between clusters, such a solution is penalized, because the clusters and its consensus sequence should represent segments of haplotypes (possibly more than one, because the if haplotypes have high local similarity, their reads might locally collapse into a single cluster). Switches are only done, if the current distribution of threads does not match the actual cluster coverages any more for a sufficiently large interval, i.e. paying the switching penalty becomes cheaper than paying the coverage penalties over and over for the entire interval.
If the coverage penalties are lowered, the threads will just stay longer on their initial clusters. However, this would still nnot incentivize the phaser to place a thread on the lowly covered haplotype, because if the only criterion is to avoid switches, it could as well place all threads on the same cluster all the time, if that is the best way to avoid switches.

**Comment by schrins** [comment]
Closing for now, as there seem to be no follow-up questions.
